### DISCUSSION TITLE
How to do a custom build?

### DISCUSSION THREAD

**Discussion by rhpijnacker** [home-assistant/operating-system#2942] [Q&A]
I'm trying to figure out why starting version 10.0.rc1 there is a problem with SwitchBot devices.
I have recently installed both 9.5 and 10.0.rc1 with the latest version of Home Assistant (in a Proxmox instance).
It is clear that with 9.5 things are ok, and 10.0.rc1 is the first version that has issues.

I would like to narrow down the exact commit that breaks it. I thought I could simply fork and build using github workflows, but that seems to fail on a missing cert.pem (https://github.com/rhpijnacker/haos/actions/runs/6961283898).
Is there a way to fix this?
Or can I do a build in some other way (e.g. a build on my local hardware)?

**Comment by agners** [comment]
Currently our Buildroot scripts and GitHub actions really require a key/certificate to create an update bundle. You can create a build without creating an update bundle locally, which means you'll have to flash the device from scratch.

There is development documentation at https://developers.home-assistant.io/docs/operating-system/getting-started which should allow you to build locally.

So I assume you have issues with Bluetooth then? There were some bugs in early 10.x versions related to the BlueZ variant. Did you try using the latest 11.1 release? (You might need to update twice, to 10.5 then to 11.1). If you are using your production setup to do testing make sure to create a full backup and download it just in case :sweat_smile:

**Comment by rhpijnacker** [comment]
I have been keeping up-to-date with Home Assistant, and Home Assistant OS within a few days of new releases.
Ever since 10.0 came out, my SwitchBot Curtain devices aren't responding like they used to.

I am using a test setup (with a different Bluetooth chip in it, actually an old iMac) with Proxmox, similar to how I run my production HA instance. I was able to reproduce exactly the same problem on this hardware.

See also https://github.com/home-assistant/core/issues/92839.
Some people have just been stuck on 9.5 because of this.
Would be a real kicker if we can fix this.

**Reply by agners** [reply]
One of the big changes with HAOS 10.0 was the new latest LTS Linux kernel 6.1 (see https://github.com/home-assistant/operating-system/releases/tag/10.0.rc1). Quite some parts of the Bluetooth stack got reworked in this new kernel release, and kernel bugs got reported immediately (see #2485 and #2535), these issues got both resolved in later HAOS releases and you should not see those in 11.1 anymore.

I fear that if you try to bisect your problem, you will simply find that the Linux kernel 6.1 is at fault. But would definitly be interesting to know if this is indeed the case.

Do you see Bluetooth errors in the host logs? Since your expirment show that this is a OS issue, I'd recommend to open an issue in the repository (https://github.com/home-assistant/operating-system/issues). Ideally with logs, and since 10.0 is a known broken version when it comes to Bluetooth I recommend to attach logs how things behave with the latest version 11.1.

**Comment by rhpijnacker** [comment]
Validating that it is indeed the commit with the kernel update is also useful information, don't you think?
If you have a suggestion on how we could trace this problem in some other way, I would be happy to provide any information needed.

When I look in homeassistant.log, I do not see any problems being reported. Did you mean for me to look in some other log file?

**Reply by agners** [reply]
> Validating that it is indeed the commit with the kernel update is also useful information, don't you think?

Definitely! I am not against your idea of bisecting the problem by rebuilding... It just is quite some effort to get it done.

> When I look in homeassistant.log, I do not see any problems being reported. Did you mean for me to look in some other log file?

When creating a new issue as suggested above the issue template will guide you through what information we typically need to pinpoint OS issues (specifically the host logs).

**Comment by avi951** [comment]
So, could it be possible to build a Home Assistant Operating system for a custom board?
Or, does it require a Docker image, like the officially supported board?
